add lstm export support

anderspapitto opened the pull request.

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -597,18 +657,18 @@ def prepare(cls, model, device='CPU', **kwargs):
 
     @classmethod
-    def _common_onnx_node_to_caffe2_op(cls, onnx_node, opset_version):
+    def _common_onnx_node_to_caffe2_op(cls, onnx_model, model_helper, onnx_node, opset_version):
```
this function should NOT take `model_helper` as there are call sites which manipulate the return values (e.g. appending the outputs). If the implementation actually appends to model_helper it'd break

anderspapitto replied:
I believe it's safe in practice. All the call sites which mutate the output do so after calling `_common_onnx_node_to_caffe2_op` on a particular node type (e.g. Slice). The only op that actually uses model_helper is LSTM, and no one tries to mutate the outputs after operating on an LSTM.

That said, yeah it's still a bit ugly. But what's the alternative? I need model_helper.

dzhulgakov replied:
I was originally suggesting to make only _common_onnx_node_to_caffe2_op not receive model_helper and every other translation function to append to model_helper directly instead of returning a list of ops.

You current approach seems to work too, so let's go with it

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -715,21 +776,26 @@ def _onnx_model_to_caffe2_net(cls, onnx_model, device, opset_version, include_in
+        # We use two separate ModelHelpers, so that the rest of the
+        # code doesn't have to worry about whether to add to
+        # param_init_net or net.
+        init_model_helper = ModelHelper(name=onnx_model.graph.name+'_init')
```
remove _init?

anderspapitto replied:
why? This matches the behavior before the diff (see the red code on the left)

dzhulgakov replied:
because the model helper will name the net 'graph.name' and 'graph.name+"_init"': https://github.com/caffe2/caffe2/blob/master/caffe2/python/model_helper.py#L118

anderspapitto replied:
++

resolved by not using ModelHelper anymore

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -297,7 +299,65 @@ def _create_gemm(cls, n, opset_version):
+
+        hidden_size = n.attrs['hidden_size']
+
+        # ad-hoc, informally-specified, bug-ridden, slow
```
it's much more plausible to infer it from the weight matrix instead. It's much more likely to be a graph initializer and will break less often. In general, yes, we need proper shape inference

anderspapitto replied:
++ I'll try switching to the weight matrix. Or maybe even both

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -297,7 +299,65 @@ def _create_gemm(cls, n, opset_version):
+        model_helper.net.Slice(B, name + "_bias_i2h",   starts=[0],             ends=[4*hidden_size])
+        model_helper.net.Slice(B, name + "_bias_gates", starts=[4*hidden_size], ends=[8*hidden_size])
+
+        # caffe2 has a different order from onnx. We need to rearrange
```
so the plan is to generate ops and rely on constant folding later? Alternative we could manipulate graph initializers explicitly too...

anderspapitto replied:
as before, unfortunately no, currently the optimization passes don't operate on Caffe2 (though we could add that).

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -297,7 +299,65 @@ def _create_gemm(cls, n, opset_version):
+
+        name = dummy_name()
+
+        # input and recurrence biases are squashed together in onnx but not in caffe2
```
all of below should go to init_net probably (assuming that the weights are there). I guess you can check that inputs to weights are part of graph initializer.

Or is the plan to run 'staging' graph pass you have?

anderspapitto replied:
unfortunately, this code runs after the optimizer, so it will not get pulled out into the init net. I could potentially put things into the init net explicitly, but that's more complicated (and, will these slices even show up as being expensive relative to the 'real' operations? I have no idea).

the pipeline is
ONNX ------(onnx-optimize)----->ONNX-------(this onnx-caffe2 code)------->Caffe2

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -297,7 +299,65 @@ def _create_gemm(cls, n, opset_version):
+            forward_only=True,
+        )
+        hidden_t_all, hidden_t_last, _, _ = lstm_out
+        model_helper.net.Copy(hidden_t_all, n.outputs[0])
```
this is somewhat hacky. You can 'rename' the blobs in the net (e.g. with net.Clone). Or at least add a TODO for it

anderspapitto replied:
yeah, this is part of dealing with the particular interface of rnn_cell.LSTM, I have no control over the output names.

I might try out Clone, though there's some annoying details there as well (do I Clone every time I process an LSTM op, or try to collect information and do it once at the end, etc.)

anderspapitto commented:
fixed tests, addressed most comments.

ModelHelper is now local to create_lstm, so it doesn't mess with the interface anymore.

I explicitly put the slice/concat operations in the init net now. It's a little ugly, and I'd prefer in the long run for this to be done by an optimization pass (possibly in onnx-optimize), but for now this works.

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -297,7 +300,71 @@ def _create_gemm(cls, n, opset_version):
+            blob_remap={ hidden_t_all: n.outputs[0], hidden_t_last: n.outputs[1] }
+        )
+
+        init_ops = [ op for op in model_helper.InitProto().op if 'Fill' not in op.type ]
```
it's a little bit sketchy. Maybe just construct your 'reform' ops in a separate net? Or do you need some ops produced by LSTMCell directly?

anderspapitto replied:
separate net could work, I'll try that

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -297,7 +300,71 @@ def _create_gemm(cls, n, opset_version):
+            model_helper.param_init_net.Concat(
+                [name_from + suffix for suffix in ("_i", "_f", "_o", "_c")],
+                [name_to, dummy_name()], axis=0)
+        reform(W, name + "/i2h_w", [0], [input_size])
```
it's not obvious at all that those names match the ones constructed by LSTMCell and might accidentally break. It seems that if you pass 'return_params=True' to rnn_cell.LSTM() it'd return you `input` and `recurrent` params that are more of a standard API. Would it be sufficient for your needs?

anderspapitto replied:
++ I'll try return_params

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -297,7 +300,71 @@ def _create_gemm(cls, n, opset_version):
+
+        # ad-hoc, informally-specified, bug-ridden, slow
+        # implementation of shape inference
+        for x in onnx_model.graph.input:
```
so what about analyzing weight shapes instead? the current thing is super sketchy and cover fewer cases as the input to LSTM layer isn't necessarily the graph input or another LSTM

anderspapitto replied:
yeah I tried briefly and it wasn't quite trivial (I think because the init-net-splitting means that the weights are not accessible in the same onnx model as the LSTM op is), but I'll try again.

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -297,7 +301,86 @@ def _create_gemm(cls, n, opset_version):
+        assert input_size is not None, "best-effort shape inference for LSTM input failed"
+
+        name = dummy_name()
+        init_mh = ModelHelper()
```
it doesn't need to be a model_helper - just core.Net() is sufficient

dzhulgakov commented on onnx_caffe2/backend.py:
```diff
@@ -597,18 +680,20 @@ def prepare(cls, model, device='CPU', **kwargs):
             translator = cls._common_onnx_node_to_caffe2_op
-        ops = translator(OnnxNode(node_def), opset_version)
+        ops = translator(init_model, pred_model, OnnxNode(node_def), opset_version)
+        if isinstance(ops, tuple):
```
maybe create a`namedtuple` for the return type? It'd be a bit safer than just returning 3 values.
